lib/Sema/ConstraintSystem.cpp

```diff
@@ -2190,9 +2190,34 @@ void ConstraintSystem::resolveOverload(ConstraintLocator *locator,
                                               openedFullType,
                                               refType};
 
+  // If we have something like '(s.foo)()', where 's.foo()' returns an IUO,
+  // then we need to only create a single constraint that binds the
+  // type to an optional.
+  auto isIUOCallWrappedInParens = [&]() {
+    auto decl = choice.getDecl();
+    auto type = decl ? decl->getInterfaceType() : nullptr;
+    if (!type || !type->is<AnyFunctionType>())
+      return false;
+
+    auto expr = locator->getAnchor();
+    if (!expr)
+      return false;
+
+    if (isa<CallExpr>(expr)) {
```

Comment:
I think these cases seem to be a different issue. None of the locators, associated with member overload set, in these examples are going to be anchored in `ParenExpr`. That said, I think we indeed need to come up with a better solution than what we currently have, creating new disjunctions in multiple distinct places is not robust enough.

Reply:
@xedin Just to clarify – these are cases that compiled before this change, and no longer compile after. The parent `ParenExpr` comes from their use as arguments in a function call, so we incorrectly consider them to be "IUO calls wrapped in parens".

Reply:
Ah, right because it's checking for parent to be a `ParentExpr`, for some reason I thought that it's checking for anchor to be a paren. Looks like we need to revert and find a better alternative.

Reply:
Those cases do not compile though.

1. cannot convert value of type 'KeyPath<S, Int?>' to expected argument type 'KeyPath<S, Int>'
2. value of optional type 'Int?' must be unwrapped to a value of type 'Int'

Reply:
@theblixguy They do compile on 5.1

Reply:
I meant they don't compile now, but yeah they did before. As you said, maybe we can change the locators to point to be anchored on the paren expression in these cases? I suppose we would still have to look for the parens and its context to figure out whether the paren is part of the call (i.e. `(s.foo()`)or if it's not (i.e. `(s.foo)()`).

Reply:
I initially found that comparing the call's `fn` and `semanticFn` gave the best results. They seemed to always be equal unless there were parens involved in other positions (like `(s.foo)()`). Maybe there's an even better way though...

Reply:
So, perhaps the check could be:

1. Is this a call, keypath or subscript expression? If yes, return false.
2. Is the fn and semanticFn equal? If yes, return false.
3. Return true

Do you find anything wrong with this logic?

Reply:
@theblixguy One potential problem with comparing `getFn()` with `getSemanticFn()` is that they won't be equal for cases like this:

```swift
func foo() -> Int! { 0 }
let x: Int = foo.self()
```

which (I think?) we want to allow. In addition, they will be equal for cases like this:

```swift
import Foundation

class C {
  @objc func foo() -> String! { "" }
}

func bar(_ x: AnyObject) {
  let str: String = (x.foo)!()
}
```

which we want to reject. I think what we'll want to do is check whether there is a `ParenExpr` between the call's `getFn()` and its "direct callee" (as determined by `markDirectCallee`).

If we turned `FunctionRefKind` into a class that stores an additional flag for "does this reference decay to a function value such that it doesn't produce an IUO?", I believe we could do this logic directly in `markDirectCallee` by setting the flag if we had to look through a paren.

Reply:
Okay, I will give that a try tomorrow 👍

Reply:
Sorry it took me a bit to get back to this today, just wanted to mention that I concur with what @hamishknight is proposing.

Reply:
```
func foo() -> Int! { 0 }
let x: Int = foo.self()
```

I'd reject this one myself, but also probably zero people are relying it, even more than the parentheses.